appinventor/blocklyeditor/src/workspace_svg.js

```diff
@@ -608,20 +608,53 @@ Blockly.WorkspaceSvg.prototype.customContextMenu = function(menuOptions) {
   /**
    * Function that returns a name to be used to sort blocks.
    * The general comparator is the block.category attribute.
-   * In the case of 'Components' the comparator is the instanceName of the component if it exists
-   * (it does not exist for generic components).
    * In the case of Procedures the comparator is the NAME(for definitions) or PROCNAME (for calls)
+   * In the case of 'Components' the comparator is the type name, instance name, then event name
    * @param {!Blockly.Block} block the block that will be compared in the sortByCategory function
    * @returns {string} text to be used in the comparison
    */
   function comparisonName(block){
-    if (block.category === 'Component' && block.instanceName)
-      return block.instanceName;
-    if (block.category === 'Procedures')
-      return (block.getFieldValue('NAME') || block.getFieldValue('PROCNAME'));
-    return block.category;
+    // Add trailing numbers to represent their sequence
+    if (block.category == 'Variables'){
+      return ('1,' + block.type);
+    }
+    if (block.category === 'Procedures'){
+      return ('2,'+ (block.getFieldValue('NAME') || block.getFieldValue('PROCNAME')));
+    }
+    if (block.category == 'Component'){
+      var component = block.type + ',' + block.instanceName + ',' + block.eventName;
```

Comment:
This doesn't take into account top-level component method and property blocks (i.e., block.eventName may be undefined). Moreover, we will be allowing "any component" event handlers in the next release, which means that block.instanceName may not always be defined.